### PR TITLE
[WIP] continue PR #784

### LINKED ISSUE #190: Cumulative gradient？Using small BatchSize to simulate big BatchSize
Could you develop the "Cumulative gradient"？
It using small BatchSize to simulate big BatchSize.

eg: [PyTorch中在反向传播前为什么要手动将梯度清零? ](https://www.zhihu.com/question/303070254/answer/573037166)

### PR DESCRIPTION
**Pull request by mzr1996**
## Motivation

This PR is the update of PR #784, to fix #190

## Modification

Based on #784, add docstring and unit tests, fix a typo, apply changes request.

## Use cases

```python
>>> # Use cumulative_iters to simulate a large batch size
>>> # It is helpful when the hardware cannot handle a large batch size.
>>> loader = DataLoader(data, batch_size=64)
>>> optimizer_hook = GradientCumulativeOptimizerHook(cumulative_iters=4)
>>> # almost equals to
>>> loader = DataLoader(data, batch_size=256)
>>> optimizer_hook = OptimizerHook()
```

To use `GradientCumulativeOptimizerHook`, modify the config to
```python
_base_ = [...]

data = dict(samples_per_gpu=16)  # Assume original samples_per_gpu=32
optimizer_config = dict(type='GradientCumulativeOptimizerHook', cumulative_iters=2)
```

### Faster RCNN (Without BatchNorm, thanks @BIGWangYuDong)
| samples_per_gpu | cumulative_iters | bbox_mAP |
|:-------------:|:---------------------:|:---------------:|
| 2               | 1                          | 0.3770        |
| 1               | 2                          | 0.3770        |
| 1               | 1                          | 0.3660        |
### ResNet-34 (With BatchNorm)
| samples_per_gpu | cumulative_iters | accuracy |
|:-------------:|:---------------------:|:---------------:|
| 32               | 1                          | 73.85        |
| 16               | 2                          | 73.25        |
| 16               | 1                          | 73.19        |
### ResNet-50 (Fp16, With BatchNorm)
| samples_per_gpu | cumulative_iters | accuracy |
|:-------------:|:---------------------:|:---------------:|
| 32               | 1                          | 76.32        |
| 16               | 2                          | 75.92        |
| 16               | 1                          | 75.81        |

## Checklist

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by zhouzaida** [comment]
please provide the performance comparison before and after using the hook

File: mmcv/runner/hooks/optimizer.py [hunk 1]
```diff
@@ -42,6 +42,94 @@ def after_train_iter(self, runner):
...
+    def has_batch_norm(self, module):
+        if isinstance(module, _BatchNorm):
+            return True
+        for m in module.children():
+            if self.has_batch_norm(m):
+                return True
+        return False
+
+    def _init(self, runner):
+        if runner.iter % self.cumulative_iters > 0:
```
Comment (zhouzaida) [inline mmcv/runner/hooks/optimizer.py:88]: `!=` may be more clear

Comment (zhouzaida) [reply mmcv/runner/hooks/optimizer.py:88]: Is it possible to resume from a `runner.iter` which meets `runner.iter % self.cumulative_iters == 0`

File: mmcv/runner/hooks/optimizer.py [hunk 2]
```diff
@@ -42,6 +42,94 @@ def after_train_iter(self, runner):
...
+        self.initialized = True
+
+    def after_train_iter(self, runner):
+        if not self.initialized:
+            self._init(runner)
+
+        if runner.iter < self.divisible_iters:
+            loss_factor = self.cumulative_iters
+        else:
+            loss_factor = self.remainder_iters
```
Comment (zhouzaida) [inline mmcv/runner/hooks/optimizer.py:115]: A test should be added here

File: mmcv/runner/hooks/optimizer.py (hunk 1 above)
Comment (mzr1996) [reply mmcv/runner/hooks/optimizer.py:88]: I'm afraid not, save_checkpoint is not controlled by the optimizer hook. Modifying `runner.iter` may cause more unexcepted results.

File: mmcv/runner/hooks/optimizer.py (hunk 2 above)
Comment (mzr1996) [reply mmcv/runner/hooks/optimizer.py:112]: I use a unit test with IterBasedRunner to cover it.

**Comment by BIGWangYuDong** [comment]
Run Faster R-CNN with:
-  batch size = 2, do not use GradientCumulativeOptimizerHook and cumulative_iters
  got 37.7% mAP
` bbox_mAP: 0.3770, bbox_mAP_50: 0.5860, bbox_mAP_75: 0.4120, bbox_mAP_s: 0.2210, bbox_mAP_m: 0.4130, bbox_mAP_l: 0.4850, bbox_mAP_copypaste: 0.377 0.586 0.412 0.221 0.413 0.485 `

-  batch size = 1, use GradientCumulativeOptimizerHook, with cumulative_iters=2
  got 37.7% mAP
` bbox_mAP: 0.3770, bbox_mAP_50: 0.5860, bbox_mAP_75: 0.4050, bbox_mAP_s: 0.2180, bbox_mAP_m: 0.4120, bbox_mAP_l: 0.4850, bbox_mAP_copypaste: 0.377 0.586 0.405 0.218 0.412 0.485 `

-  batch size = 1, use GradientCumulativeOptimizerHook, do not use cumulative_iters
  got 36.6% mAP
` bbox_mAP: 0.3660, bbox_mAP_50: 0.5700, bbox_mAP_75: 0.3990, bbox_mAP_s: 0.2080, bbox_mAP_m: 0.4010, bbox_mAP_l: 0.4690, bbox_mAP_copypaste: 0.366 0.570 0.399 0.208 0.401 0.469`

**Comment by mzr1996** [comment]
> please provide the performance comparison before and after using the hook

Done, updated in comments.

**Comment by zhouzaida** [comment]
hi, have you tested the `GradientCumulativeFp16OptimizerHook` in mmcls or mmdet? BTW, maybe we need to add more unittests to cover the code.

**Comment by mzr1996** [comment]
> hi, have you tested the `GradientCumulativeFp16OptimizerHook` in mmcls or mmdet? BTW, maybe we need to add more unittest to cover the code.

I have updated the test results in the comment.

And I have added unit tests that covered the fp16 related hook, but codecov doesn't recognize them. I don't know why.

**Comment by zhouzaida** [comment]
> > hi, have you tested the `GradientCumulativeFp16OptimizerHook` in mmcls or mmdet? BTW, maybe we need to add more unittest to cover the code.
> 
> I have updated the test results in the comment.
> 
> And I have added unit tests that covered the fp16 related hook, but codecov doesn't recognize them. I don't know why.
It perhaps skips the test. So do you test it in your local environment?
```
@pytest.mark.skipif(
    not torch.cuda.is_available(), reason='requires CUDA support')
```

**Comment by mzr1996** [comment]
> > > hi, have you tested the `GradientCumulativeFp16OptimizerHook` in mmcls or mmdet? BTW, maybe we need to add more unittest to cover the code.
> > 
> > 
> > I have updated the test results in the comment.
> > And I have added unit tests that covered the fp16 related hook, but codecov doesn't recognize them. I don't know why.
> > It perhaps skips the test. So do you test it in your local environment?
> 
> ```
> @pytest.mark.skipif(
>     not torch.cuda.is_available(), reason='requires CUDA support')
> ```

Local test results:
深度截图_选择区域_20210817154659

**Review by zhouzaida — APPROVED** [review]
LGTM

**Review by ZwwWayne — APPROVED** [review]

**Comment by WitcherEnvoy** [comment]
Did each experiment use the same learning rate?

**Comment by mzr1996** [comment]
> Did each experiment use the same learning rate?

Yes, since we use `GradientCumulativeOptimizerHook` to simulate the same batch size, they share the same learning rate.

**Comment by WitcherEnvoy** [comment]
> > Did each experiment use the same learning rate?
> 
> Yes, since we use `GradientCumulativeOptimizerHook` to simulate the same batch size, they share the same learning rate.

Including experiments that set cumulative_iters=1? Shouldn't the learning rate be reduced to half?

**Comment by WitcherEnvoy** [comment]
> > Did each experiment use the same learning rate?
> 
> Yes, since we use `GradientCumulativeOptimizerHook` to simulate the same batch size, they share the same learning rate.

Hi, can you answer my earlier question?

**Comment by mzr1996** [comment]
> > > Did each experiment use the same learning rate?
> > 
> > 
> > Yes, since we use `GradientCumulativeOptimizerHook` to simulate the same batch size, they share the same learning rate.
> 
> Hi, can you answer my earlier question?

Right, usually we need to reduce the learning rate when reducing the batch size, but these experiments are just used to check if `GradientCumulativeOptimizerHook` works. Maybe more experiments are needed to test the effectiveness of gradient accumulating.
